common4j/src/main/com/microsoft/identity/common/java/controllers/ExceptionAdapter.java

```diff
@@ -436,11 +440,27 @@ public static ClientException clientExceptionFromException(@NonNull final Throwa
 
 
         if (e instanceof OutOfMemoryError) {
-            return new ClientException(
-                    ClientException.OUT_OF_MEMORY,
-                    e.getMessage(),
-                    e
-            );
+            // When receiving an out of memory error, instead of gracefully returning a failure result, we should shut down
+            // current broker process, to allow a new broker process to be launched at the next request from client app.
+            // This will result in a new broker process with fresh memory allocation, rather than repeating out of memory
+            // errors. In the case that this is an msal-only scenario, and this code is running inside client app, then
+            // client app will be shut down and will need to be launched again.
+
+            // Log status code and record exception in span
+            final Span currentSpan = SpanExtension.current();
+            currentSpan.setStatus(StatusCode.ERROR);
+            currentSpan.recordException(e);
+
+            // Attach the stack trace, to debug in telemetry later
+            currentSpan.setAttribute(AttributeName.out_of_memory_exception_stacktrace.name(),
+                    StringUtil.getStacktraceAsStringFromElementArray(e.getStackTrace()));
+
+            Logger.error(TAG, "Received an out of memory error, shutting broker process so a new one can be launched.", e);
+
+            // Shut down current process
+            // Calling client app will receive "killed unexpectedly" error from MSAL if broker is shut down with this statement
+            // Calling application should not crash, and should be able to make another call, which will result in a new broker process.
+            System.exit(0);
```

Comment:
Would this impact OOM errors thrown from broker process during WPJ?

Reply:
If you are sure that this would have minimal impact on non teams devices scenarios, I think we are good if you kill the process. But if this is also happening in large numbers, it would lead to crashes if OOM exception is thrown in broker during some broker specific tasks like WPJ or flights fetch or active broker selection process. And broker apps track crash analytics more seriously. 
You may also want to think if you need a flight.

Reply:
The issue is not exclusive to teams devices. The particular icm we got that pointed it out was from teams devices, but it's a fairly standard exception to see on android from what i can tell. I'll start a teams chat to discuss with you and Mohit

Reply:
One thing to note is that this does not cause the calling app to crash. MSAL will send an client exception gracefully to calling app, "Activity killed unexpectedly". Calling app is able to handle this exception, and either retry the request or proceed in some other way.

Reply:
We can discuss this point in scrum

Reply:
Decided to validate this with broker hosting app teams, and will instead merge this fix in feb

Reply:
Resolving as i'm no longer crashing broker